### PR TITLE
*: cancel required leader streams when memeber lost its leader

### PR DESCRIPTION
**Pull request by xiang90**
/cc @heyitsanthony @gyuho

### CODE REVIEW

File: etcdserver/api/v3rpc/interceptor.go [hunk 1]
```diff
@@ -75,3 +112,52 @@ func splitMethodName(fullMethodName string) (string, string) {
...
+				if s.Leader() == types.ID(raft.None) {
+					noLeaderCnt++
+				} else {
+					noLeaderCnt = 0
+				}
+
+				// We are more conservative on canceling existing streams. Reconnecting streams
+				// cost much more than just rejecting new requests. So we wait until the member
+				// cannot find a leader for 3 election timeouts to cancel existing streams.
+				if noLeaderCnt >= 3 {
```
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/interceptor.go]: declare election retry max as a const somewhere?

Comment (xiang90) [inline etcdserver/api/v3rpc/interceptor.go]: will do.

**Comment by gyuho** [comment]
LGTM. Defer to @heyitsanthony 

Thanks.

File: etcdserver/api/v3rpc/watch.go [hunk 2]
```diff
@@ -108,69 +111,88 @@ func (ws *watchServer) Watch(stream pb.Watch_WatchServer) error {
...
-			}
-			wsrev := sws.watchStream.Rev()
-			rev := creq.StartRevision
-			if rev == 0 {
-				rev = wsrev + 1
-			}
-			id := sws.watchStream.Watch(creq.Key, creq.RangeEnd, rev)
-			if id != -1 && creq.ProgressNotify {
-				sws.progress[id] = true
+	go func() {
```
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/watch.go]: can `recvLoop` stay the same and instead move the errc/goroutine machinery into `Watch`?

File: integration/v3_grpc_test.go [hunk 3]
```diff
@@ -1004,3 +1006,46 @@ func TestGRPCRequireLeader(t *testing.T) {
...
+	// new stream should also be rejected
+	wStream, err = wAPI.Watch(ctx)
+	if err != nil {
+		t.Fatalf("wAPI.Watch error: %v", err)
+	}
+	_, err = wStream.Recv()
+	if grpc.ErrorDesc(err) != rpctypes.ErrNoLeader.Error() {
+		t.Errorf("err = %v, want %v", err, rpctypes.ErrNoLeader)
+	}
+}
```
Comment (heyitsanthony) [inline integration/v3_grpc_test.go]: this should probably also test that streams work again once there's a new leader

File: etcdserver/api/v3rpc/watch.go (hunk 2 above)
Comment (xiang90) [inline etcdserver/api/v3rpc/watch.go]: yes. i feel this makes the driver func clearer. but i will try the other way.

File: integration/v3_grpc_test.go (hunk 3 above)
Comment (xiang90) [inline integration/v3_grpc_test.go]: ok. sure.

**Comment by heyitsanthony** [comment]
overall looks ok

**Comment by xiang90** [comment]
@heyitsanthony All fixed. PTAL.

**Comment by heyitsanthony** [comment]
lgtm

**Pull request merged**